src/utilities/time.ts

```diff
@@ -1,12 +1,171 @@
+ */
+export function isTimeBalanced(time: string, type: TimeTypes): boolean {
+  switch (type) {
+    case TimeTypes.ABSOLUTE: {
+      const balancedTime = parseDoyOrYmdTime(getDoyTime(new Date(getUnixEpochTime(time))));
+      const originalTime = parseDoyOrYmdTime(time);
+      if (balancedTime === null || originalTime === null) {
+        return false;
+      }
+      return originalTime.year === balancedTime.year;
+    }
+    case TimeTypes.EPOCH:
+    case TimeTypes.RELATIVE: {
+      const originalTime = parseDurationString(time);
+      const balancedTime = parseDurationString(getBalancedDuration(time));
+
+      if (balancedTime === null || originalTime === null) {
+        return false;
+      }
+      return (
+        balancedTime.days === originalTime.days &&
+        balancedTime.hours === originalTime.hours &&
+        balancedTime.minutes === originalTime.minutes &&
+        balancedTime.seconds === originalTime.seconds &&
+        balancedTime.milliseconds === originalTime.milliseconds
+      );
+    }
+    default:
+      return false;
+  }
+}
+
+/**
+ * Parse a duration string into a parsed duration object.
+ * If no unit is specified, it defaults to microseconds.
+ *
+ * @example
+ * parseDurationString('1h 30m');
+ * // => {
+ * // =>   days: 0,
+ * // =>   hours: 1,
+ * // =>   isNegative: false,
+ * // =>   microseconds: 0,
+ * // =>   milliseconds: 0,
+ * // =>   minutes: 30,
+ * // =>   seconds: 0,
+ * // =>   years: 0,
+ * // => }
+ * @example
+ * parseDurationString('-002T00:45:00.010')
+ * // => {
+ * // =>   days: 2,
+ * // =>   hours: 0,
+ * // =>   isNegative: true,
+ * // =>   microseconds: 0,
+ * // =>   milliseconds: 10,
+ * // =>   minutes: 45,
+ * // =>   seconds: 0,
+ * // =>   years: 0,
+ * // => }
+ * @example
+ * parseDurationString('90')
+ * // => {
+ * // =>   minutes: 0,
+ * // =>   seconds: 0,
+ * // =>   microseconds: 90,
+ * // =>   milliseconds: 0,
+ * // =>   days: 0,
+ * // =>   hours: 0,
+ * // =>   isNegative: false,
+ * // =>   years: 0,
+ * // => }
+ * @example
+ * parseDurationString('-123.456s', 'microseconds')
+ * // => {
+ * // =>   microseconds: 0,
+ * // =>   milliseconds: 456,
+ * // =>   seconds: -123,
+ * // =>   minutes: 0,
+ * // =>   hours: 0,
+ * // =>   days: 0,
+ * // =>   isNegative: true,
+ * // =>   years: 0,
+ * // => }
+ *
+ * @param {string} durationString - The duration string to parse.
+ * @param {'years' | 'days' | 'hours' | 'minutes' | 'seconds' | 'milliseconds' | 'microseconds'} units - The units to parse the duration string in.
+ * @return {ParsedDurationString} The parsed duration object.
+ * @throws {Error} If the duration string is invalid.
+ */
+export function parseDurationString(
+  durationString: string,
+  units: 'days' | 'hours' | 'minutes' | 'seconds' | 'milliseconds' | 'microseconds' = 'microseconds',
+): ParsedDurationString | never {
+  const validNegationRegex = `((?<isNegative>-))?`;
+  const validDurationValueRegex = `([+-]?)(\\d+)(\\.\\d+)?`;
```

Comment:
Do you mind using named regex groups instead of relying on the match index when extracting the results just to make the regex a little easier to read and debug?

Reply:
Yep, I added groups to the regexes